src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/haflows/HaFlowPingSpec.groovy

```diff
@@ -0,0 +1,212 @@
+                                   .forward(expectedResponseSubflowPart)
+                                   .reverse(expectedResponseSubflowPart)
+                                   .build()])
+                .build()
+
+        when: "ping HA-flow"
+        def response = northboundV2.pingHaFlow(haFlow.getHaFlowId(), new HaFlowPingPayload(2000))
+        response = 'replace unpredictable latency values from ping response'(response)
+
+        then: "HA-flow ping is not successful, but one of subflows ping is successful"
+        response == expectedResponse
+
+        cleanup:
+        Wrappers.silent{haFlowHelper.deleteHaFlow(haFlow.getHaFlowId())}
+    }
+
+    def getPacketCountOfVlanPingRule(SwitchId switchId, String encapsulationType) {
+        def pingRuleCookie = getPingRuleCookie(encapsulationType)
+        return northbound.getSwitchRules(switchId).flowEntries
+                .findAll { it.cookie == pingRuleCookie }[0].packetCount
+    }
+
+    private long getPingRuleCookie(String encapsulationType) {
+        if (FlowEncapsulationType.TRANSIT_VLAN.toString().equalsIgnoreCase(encapsulationType)) {
+            return Cookie.VERIFICATION_UNICAST_RULE_COOKIE
+        } else if (FlowEncapsulationType.VXLAN.toString().equalsIgnoreCase(encapsulationType)) {
+            return Cookie.VERIFICATION_UNICAST_VXLAN_RULE_COOKIE
+        } else {
+            throw new IllegalArgumentException("Unknown encapsulation " + encapsulationType)
+        }
+    }
+
+    def 'replace unpredictable latency values from ping response'(HaFlowPingResult originalResponse) {
+        //TODO: implement PingResponse model in test package to safely compare expected and actual responses without
+        //manipulating original response
+        def subFlowPingPayloadWithZeroLatency = originalResponse.subFlows[0].tap {
+            it.forward.latency = 0
+            it.reverse.latency = 0
+        }
+        originalResponse.subFlows[0] = subFlowPingPayloadWithZeroLatency
+        return originalResponse
+    }
+
+    private void pingRuleCountersAreGrow(SwitchTriplet swT, HaFlow haFlow) {
+        def sharedSwitchPacketCount = getPacketCountOfVlanPingRule(swT.shared.dpId, haFlow.encapsulationType)
+        def ep1SwitchPacketCount = getPacketCountOfVlanPingRule(swT.ep1.dpId, haFlow.encapsulationType)
+        def ep2SwitchPacketCount = getPacketCountOfVlanPingRule(swT.ep2.dpId, haFlow.encapsulationType)
+
+        wait(pingInterval + STATS_LOGGING_TIMEOUT + WAIT_OFFSET) {
+            assert getPacketCountOfVlanPingRule(swT.shared.dpId, haFlow.encapsulationType) > sharedSwitchPacketCount
+            assert getPacketCountOfVlanPingRule(swT.ep1.dpId, haFlow.encapsulationType) > ep1SwitchPacketCount
+            assert getPacketCountOfVlanPingRule(swT.ep2.dpId, haFlow.encapsulationType) > ep2SwitchPacketCount
+        }
+    }
+
+    private void pingRuleCountersDoesntGrow(SwitchTriplet swT, HaFlow haFlow) {
```

Comment:
Actually, you don't need this method. Just return Boolean in `arePingRuleCountersGrow()` and verify it in `then` section of the test.

Reply:
Yes. This is a common problem when you try to assert several things at once without ExpectedSomething... class.